BUG: masked array division should ignore all FPEs in mask calculation

carlosilva10260 opened the pull request:
This PR fixes the bug #25810 where an internal multiplication underflow occured with np.seterr(under="raise")

seberg commented on numpy/ma/tests/test_core.py:
```diff
@@ -2581,6 +2581,14 @@ def test_no_masked_nan_warnings(self):
             allclose(m, 0.5)
 
+    def test_masked_array_underflow(self):
+        np.seterr(under="raise")
```
Please make sure to use `with np.errstate()`, this will otherwise affect later tests randomly.  Otherwise, LGTM, no need to insert those newlines, but they seem fine.

carlosilva10260 commented:
Hi @seberg , I've fixed the test as you requested and also removed the extra spaces that I forgot to remove before.
Is everything now correct to you?

eendebakpt commented on numpy/ma/tests/test_core.py:
```diff
@@ -2581,6 +2581,14 @@ def test_no_masked_nan_warnings(self):
+            X = np.ma.array(x)
+            x2 = x/2.0  # <- works
+            X2 = X/2.0
+            assert(X2.all() == X.all() / 2.0)
```
What does this assert test?

carlosilva10260 replied:
The purpose is to make sure that the division occured fine without the multiply underflow error

eendebakpt commented on numpy/ma/tests/test_core.py:
```diff
@@ -2581,6 +2581,14 @@ def test_no_masked_nan_warnings(self):
 
+    def test_masked_array_underflow(self):
+        with(np.errstate(under="raise")):
+            x = np.arange(0, 3, 0.1)
```
Maybe move this line and the next two or of the with context, as the test is about the masked array.

carlosilva10260 replied:
Ok i'll do so

eendebakpt commented on numpy/ma/tests/test_core.py:
```diff
@@ -2581,6 +2581,13 @@ def test_no_masked_nan_warnings(self):
+        X = np.ma.array(x)
+        with(np.errstate(under="raise")):
+            X2 = X/2.0
+            assert(X2.all() == X.all() / 2.0)
```
```suggestion
        np.testing.assert_array_equal(X2, x/2)  
```
The `X.all()` and `X2.all()` are both False. Perhaps you wanted to write `assert np.all(X2==x/2.0)`?

carlosilva10260 replied:
Thank you for the correction, at the time I wasn't 100% sure how I would test it but I'll change to that

carlosilva10260 committed:
BUG: masked array division broken with np.seterr(under=raise) #25810

carlosilva10260 commented:
Hi @seberg  @eendebakpt, is everything correct now in the new PR?

seberg reviewed (approved):
LGTM, thanks.

seberg committed:
STY: Remove unnecessary brackets
